src/index.js

```diff
@@ -77,10 +78,14 @@ class CopyPlugin {
                   return Promise.resolve();
                 }
 
+                const limit = pLimit(globalRef.concurrency || Infinity);
```

Comment:
We already have this https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/processPattern.js#L25, maybe we should move it on top

Reply:
I think we should keep it here:
```
const limit = pLimit(globalRef.concurrency || Infinity);
return Promise.all(
  files
    .filter(Boolean)
    .map((file) =>
      limit(() => postProcessPattern(globalRef, pattern, file))
    )
);
```
and remove  [ https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/processPattern.js#L25](https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/processPattern.js#L25)
Because the following code is a synchronous process and i don't understand why it's used here.
I'm confused.
```
        limit(() => {
          const file = {
            force: pattern.force,
            absoluteFrom: path.resolve(pattern.context, from),
          };

          file.relativeFrom = path.relative(pattern.context, file.absoluteFrom);

          if (pattern.flatten) {
            file.relativeFrom = path.basename(file.relativeFrom);
          }

          logger.debug(`found ${from}`);

          // Check the ignore list
          let il = pattern.ignore.length;

          // eslint-disable-next-line no-plusplus
          while (il--) {
            const ignoreGlob = pattern.ignore[il];

            let globParams = {
              dot: true,
              matchBase: true,
            };

            let glob;

            if (typeof ignoreGlob === 'string') {
              glob = ignoreGlob;
            } else if (isObject(ignoreGlob)) {
              glob = ignoreGlob.glob || '';

              const ignoreGlobParams = Object.assign({}, ignoreGlob);
              delete ignoreGlobParams.glob;

              // Overwrite minimatch defaults
              globParams = Object.assign(globParams, ignoreGlobParams);
            } else {
              glob = '';
            }

            logger.debug(`testing ${glob} against ${file.relativeFrom}`);

            if (minimatch(file.relativeFrom, glob, globParams)) {
              logger.info(
                `ignoring '${file.relativeFrom}', because it matches the ignore glob '${glob}'`
              );

              return Promise.resolve();
            }

            logger.debug(`${glob} doesn't match ${file.relativeFrom}`);
          }

          // Change the to path to be relative for webpack
          if (pattern.toType === 'dir') {
            file.webpackTo = path.join(pattern.to, file.relativeFrom);
          } else if (pattern.toType === 'file') {
            file.webpackTo = pattern.to || file.relativeFrom;
          } else if (pattern.toType === 'template') {
            file.webpackTo = pattern.to;
            file.webpackToRegExp = pattern.test;
          }

          if (path.isAbsolute(file.webpackTo)) {
            if (output === '/') {
              const message =
                'using older versions of webpack-dev-server, devServer.outputPath must be defined to write to absolute paths';

              logger.error(message);

              compilation.errors.push(new Error(message));
            }

            file.webpackTo = path.relative(output, file.webpackTo);
          }

          logger.info(
            `determined that '${from}' should write to '${file.webpackTo}'`
          );

          return file;
        })
```

Reply:
Feel free to do it

Reply:
> Feel free to do it

ok.